service/frontend/adminHandler.go

```
@@ -156,47 +160,127 @@ func (adh *AdminHandler) AddSearchAttribute(ctx context.Context, request *admins
+	resp, err := adh.getSearchAttributes(ctx, indexName, run.GetRunID())
+	if err != nil {
+		// Don't return error in case of getSearchAttributes error. Just log it.
+		adh.GetLogger().Warn("Unable to get search attributes back while adding them.", tag.Error(err))
+		metricsGetSearchAttributesScope := adh.GetMetricsClient().Scope(metrics.AdminGetSearchAttributesScope)
+		metricsGetSearchAttributesScope.IncCounter(metrics.ServiceFailures)
+	}
+
+	return &adminservice.AddSearchAttributesResponse{
+		CustomAttributes:         resp.GetCustomAttributes(),
+		SystemAttributes:         resp.GetSystemAttributes(),
+		Mapping:                  resp.GetMapping(),
+		AddWorkflowExecutionInfo: resp.GetAddWorkflowExecutionInfo(),
+	}, nil
+}
+
+func (adh *AdminHandler) GetSearchAttributes(ctx context.Context, request *adminservice.GetSearchAttributesRequest) (_ *adminservice.GetSearchAttributesResponse, retError error) {
+	defer log.CapturePanic(adh.GetLogger(), &retError)
+
+	scope, sw := adh.startRequestProfile(metrics.AdminGetSearchAttributesScope)
+	defer sw.Stop()
+
+	indexName := request.GetIndexName()
+	if indexName == "" {
+		indexName = adh.ESConfig.GetVisibilityIndex()
+	}
+
+	resp, err := adh.getSearchAttributes(ctx, indexName, "")
+	if err != nil {
+		return nil, adh.error(err, scope)
+	}
+
+	return resp, nil
+}
+
+func (adh *AdminHandler) getSearchAttributes(ctx context.Context, indexName, runID string) (*adminservice.GetSearchAttributesResponse, error) {
+	var lastErr error
+	descResp, err := adh.GetSDKClient().DescribeWorkflowExecution(ctx, addsearchattributes.WorkflowName, runID)
+	var wfInfo *workflowpb.WorkflowExecutionInfo
+	if err != nil {
+		// NotFound can happen when no search attributes were added and the workflow has never been executed.
+		if _, notFound := err.(*serviceerror.NotFound); !notFound {
+			lastErr = serviceerror.NewInternal(fmt.Sprintf("unable to get %s workflow state: %v", addsearchattributes.WorkflowName, err))
```

why not return earlier?

lastErr will also be overrided later

---

So this `getSearchAttributes` is called from two places `GetSearchAttributes` and `AddSearchAttributes`. For the second one, I decided not to fail entire call if get has failed (primary goal for this API is to add search attribute, not to list it). But I still want to collect as much info as I can (call to ES mapping and SDK client are independent). From other side `GetSearchAtributes` should fail if some data is not available. I guess it doesn't matter if it fails with first error or with last one.

---

i would prefer not shadowing the error

if an error is not important, then log it then ignore.
if an error is important, then return it.

---

I reordered calls in the right priority order and added logging for every error.